### PR TITLE
fix: Collect multiple smartrest messages from c8y devicecontrol topic

### PR DESCRIPTION
**Pull request by Bravo555**
## Proposed changes

c8y/devicecontrol/notifications topic, as many others, can contain in its smartrest payload multiple records. For now ensure that multiple records are parsed from this topic in particular, as C8y sends multiple operations in a single message if e.g. device is offline, multiple operations are started, and then the device connects, it then receives all those operations in a single MQTT message.

But a longer term solution would be to refactor the code to ensure that we're supporting collecting multiple smartrest messages from all the possible MQTT topics.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

#3297

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 1]
```diff
@@ -632,24 +632,36 @@ impl CumulocityConverter {
         &mut self,
         message: &MqttMessage,
     ) -> Result<Vec<MqttMessage>, ConversionError> {
-        let operation = C8yOperation::from_json(message.payload.as_str()?)?;
-        let device_xid = operation.external_source.external_id;
-        let cmd_id = self.command_id.new_id_with_str(&operation.op_id);
+        // Smartrest messages on c8y/devicecontrol/notifications can contain multiple operations in a single MQTT
+        // message, so split them
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:636]: Terminoligy correction. The messages arrving on `c8y/devicecontrol/notifications` is **JSON over MQTT** messages. **SmartREST** is the comma separated csv-like messages.

```suggestion
        // JSON over MQTT messages on c8y/devicecontrol/notifications can contain multiple operations in a single MQTT
        // message, so split them
```

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 2]
```diff
@@ -632,24 +632,36 @@ impl CumulocityConverter {
...
+        let operation_payloads = collect_smartrest_messages(message.payload_str()?);
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:637]: I was surprised that `collect_smartrest_messages()` works here. However, after seeing the contents of the function, it makes sense. That means, the function should be renamed now, otherwise it's very misleading. Maybe `collect_c8y_messages()` or so on.

**Review by rina23q — COMMENTED** [review]
Some renaming required.

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 2 above)
Comment (Bravo555) [reply crates/extensions/c8y_mapper_ext/src/converter.rs]: Right, it's not Smartrest, so probably it would also be fine to just split on `\n`, right? As JSON over MQTT doesn't have these escaping problems the function was originally made for?

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 3]
```diff
@@ -632,24 +632,36 @@ impl CumulocityConverter {
...
+                    operation.op_id.clone(),
+                    &operation.extras,
+                    &operation_message,
+                )
+                .await;
+            let result = self.handle_c8y_operation_result(&result, Some(operation.op_id));
+            output.extend(result);
+        }
 
         Ok(output)
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:666]: My opinion for the future development.

What I don't like from the current way (due to the converter design) is, this return value Ok(output) is a vector of MQTT messages that will be published later at the same time. If 5 operations are included in one message from c8y, 5 of 504 messages will be sent together. But it's definitely out of scope to fix now.

**Review by rina23q — APPROVED** [review]
LGTM

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 3 above)
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:666]: > What I don't like from the current way (due to the converter design) is, this return value Ok(output) is a vector of MQTT messages that will be published later at the same time.  If 5 operations are included in one message from c8y, 5 of 504 messages will be sent together.

What don't you like in the current design? That the value returned by this method is a vector of MQTT messages or that the five 504 messages are not packed into a single MQTT message? The later can be implemented using the former method signature.

**Review by didier-wenzek — APPROVED** [review]
Approved

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 3 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:666]: `504` changes an operation state to executing in c8y.

Let's say there are 3 operation (Op_A, Op_B, Op_C) are delivered at the same.
Ideally,
1. `Op_A` start proceeding -> Publish `504` for Op_A
2. `Op_B` start proceeding -> Publish `504` for Op_B
3. `Op_C` start proceeding -> Publish `504` for Op_C

However, with the current design, when all of `Op_A`, `Op_B`, and `Op_C` start proceeding, mapper publishes three `504` messages for these 3 operations. 

In short, what I don't like is the timing the `504` message is published by mapper. Not critical, but minor point.

Comment (reubenmiller) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:666]: Though this doesn't occur for operations being processed by a workflow right (as the workflow execution is controlled by the tedge-agent, and currently only processes one type of operation at a time of the same type).
